### PR TITLE
s390x hw support for aes-ecb/ofb/cfb/cfb8

### PR DESCRIPTION
**Pull request by p-steuer**
Add s390 hw support for various aes modes. I put those in one PR since its not so much code per mode: Each patch just adds assembly routine for the corresponding instruction plus some code for partial block processing.

### CODE REVIEW

**Comment by p-steuer** [comment]
rebased this because of ccm merge

**Comment by richsalz** [comment]
pign @dot-asm

**Comment by dot-asm** [comment]
What I don't appreciate about this is that decision about whether or not to engage hardware is taken twice. Or rather it relies on second check to have same outcome. I mean you check for capability bit, and return a EVP_CIPHER structure with pointer to common AES_set_encrypt_key and hardware-specific cipher subroutine. And trouble is that AES_set_encrypt_key checks for capability too, and if doesn't match at that point, bad things will happen. Even though outcome of the check is normally same, and probability of failure [cause by data corruption] is diminishingly low, the approach is kind of not quite right. The data structure should be self-consistent, in sense that it would be more appropriate if structure would point at key setup procedure that sets up key suitable for use by hardware without checking for capability bit.

Just in case for others' reference. The way it worked so far [in modes in question] was that AES_set_encrypt_key was checking the processor capabilities initiated key schedule differently depending on the outcome. And then AES_encrypt was observing how key schedule looks like and acted accordingly. So that check for hardware capability was done at key initialization, and once initialized, AES_encrypt couldn't be knocked off track.

**Comment by dot-asm** [comment]
Regarding my last comment. One can also rephrase it as following. In long run it's probably appropriate to omit the check for hardware capability in AES_set_encrypt_key. And with this in mind modifications that bring us closer to this goal would be viewed as preferred over those that bring us further. Key question if we can agree that it's meaningful goal...

**Comment by p-steuer** [comment]
>  @dot-asm : I mean you check for capability bit, and return a EVP_CIPHER structure with pointer to common AES_set_encrypt_key and hardware-specific cipher subroutine. [...] The data structure should be self-consistent, in sense that it would be more appropriate if structure would point at key setup procedure that sets up key suitable for use by hardware without checking for capability bit.

I reworked this to set the pointer to a hardware-specific init routine instead of AES_set_encrypt_key i.e., it now works the same as in case of gcm/ccm.

Another advantage is that you save the copy / cleanse of the param block on each cipher operation.

**Comment by p-steuer** [comment]
> In long run it's probably appropriate to omit the check for hardware capability in AES_set_encrypt_key. [...] Key question if we can agree that it's meaningful goal...

.. so the old non-evp apis wouldnt have hw support ? I agree its a meaningful goal to move as much of the capability checking as possible out of the asm. At the moment there are applications which still use the non-evp apis, so its nice to still have basic hw support there.

**Comment by richsalz** [comment]
Yes, it's common (AES on x86) that only the EVP interface has hardware support.  Think of it as motivation to move to EVP, which is our preferred API.  But there is nothing that says the legacy API's *cannot* support hardware ...

**Comment by dot-asm** [comment]
Just to be on the safe side, could you re-base this [to current master top]?

**Comment by p-steuer** [comment]
Rebased. I ll squash the update commit(s) appropriately once this is approved.

File: crypto/evp/e_aes.c [hunk 1]
```diff
@@ -1076,54 +1125,223 @@ static int s390x_aes_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
...
+            cctx->kmf.param.cv[n] = enc ? out[n] : tmp;
+            ++n;
+        }
+    }
+
+    cctx->res = n;
+    return 1;
+}
+
+# define S390X_aes_128_cfb8_CAPABLE (S390X_aes_128_CAPABLE &&		\
```
Comment (p-steuer) [inline crypto/evp/e_aes.c]: Remove the S390X_aes_xxx_CAPABLE checks, technically we dont need km in this case.

Comment (dot-asm) [reply crypto/evp/e_aes.c]: I don't follow. Is it question? Suggestion? Assertion? Something you haven't pushed yet?

Comment (p-steuer) [reply crypto/evp/e_aes.c]: Noticed after the rework which incorporated your feedback. Just a note to myself.

Comment (dot-asm) [reply crypto/evp/e_aes.c]: This however calls for following question. If we aim for removal of capability detection in AES_set_encrypt_key, then modes in question (as well as some other) will drop slooooow on system that doesn't have kmf (or whichever km extension). Something to keep in mind...

Comment (dot-asm) [reply crypto/evp/e_aes.c]: Just in case to clarify "removal of capability detection in AES_set_encrypt_key" does not mean "here and now", but rather "in some distant future".

File: crypto/evp/e_aes.c [hunk 2]
```diff
@@ -1076,54 +1125,223 @@ static int s390x_aes_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
...
+        *out = *in ^ cctx->kmo.param.cv[n];
+        n = (n + 1) & 0xf;
+        --len;
+        ++in;
+        ++out;
+    }
+
+    rem = len & 0xf;
+
+    len &= ~0xf;
```
Comment (dot-asm) [inline crypto/evp/e_aes.c]: This kind of thing always makes me nervous. Is sign extended or not? I mean sequence of how compiler interprets this in 64-bit case is a) take 32-bit *signed* positive integer 0xf; b) invert bitwise, still 32-bit signed but negative; b) left hand is unsigned 64-bit, so *extend* sign to upper 32 bits. However, what makes me nervous is not that compiler would fail to extend the sign bit, but rather what *we* think is going on when seeing/writing this. I mean for example did you, Patrick, explicitly thought of this sign extension, or was just semi-automatic it-knows-what-I-mean? Do note that I'm not saying that you didn't think this, or that the code is wrong, question effectively goes to everybody. Does one actually/always recognize the implicit sign extension when one reads it?

Now, this is a side note (i.e. no action is expected), but one can wonder if making it more explicit would improve readability? And if so, if it would appropriate to change this [to ~(size_t)0xf]? Thoughts?

Comment (p-steuer) [reply crypto/evp/e_aes.c]: I went through this thought-process (that this works for quotient/remainder with size_t divided by 16) once, when i noticed compiler would not optimize len -= rem in this case, and since then use this as an idiom.

I ll add the cast for i agree it helps readability.

File: crypto/evp/e_aes.c (hunk 1 above)
Comment (p-steuer) [reply crypto/evp/e_aes.c]: > This however calls for following question. [...]

I dont see how question is related to my note, but in general: I agree to leave this in the distant future i.e, only to have this goal in mind when adding new features. Its essantially "cleaner code" vs "slow down old apps and new apps on old machines" and the latter only gets better in time.

File: crypto/evp/e_aes.c [hunk 3]
```diff
@@ -1076,54 +1125,220 @@ static int s390x_aes_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
...
+    const int fc_flags = enc ? 0 : S390X_DECRYPT;
 
-# define S390X_aes_128_ofb_CAPABLE	0
-# define S390X_aes_192_ofb_CAPABLE	0
-# define S390X_aes_256_ofb_CAPABLE	0
+    s390x_km(in, len, out, cctx->fc | fc_flags, &cctx->km.param);
+    return 1;
+}
+
+# define S390X_aes_128_ofb_CAPABLE (S390X_aes_128_CAPABLE &&		\
```
Comment (dot-asm) [inline crypto/evp/e_aes.c]: Well, if you omitted \*_cfb8_CAPABLE, then why keep these [and \*_cfb_CAPABLE]?

File: crypto/evp/e_aes.c [hunk 4]
```diff
@@ -1076,54 +1125,220 @@ static int s390x_aes_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
...
+}
 
-# define s390x_aes_ecb_cipher aes_ecb_cipher
 static int s390x_aes_ecb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
-                                const unsigned char *in, size_t len);
+                                const unsigned char *in, size_t len)
+{
+    S390X_AES_ECB_CTX *cctx = EVP_C_DATA(S390X_AES_ECB_CTX, ctx);
+    const int enc = EVP_CIPHER_CTX_encrypting(ctx);
+    const int fc_flags = enc ? 0 : S390X_DECRYPT;
```
Comment (dot-asm) [inline crypto/evp/e_aes.c]: One can set this flag at key setup time.

File: crypto/evp/e_aes.c (hunk 3 above)
Comment (p-steuer) [reply crypto/evp/e_aes.c]: I omitted S390X_aes_xxx_CAPABLE in case of cfb8 only, because cfb8 does not need km function code to process partial block.

File: crypto/evp/e_aes.c (hunk 4 above)
Comment (p-steuer) [reply crypto/evp/e_aes.c]: I moved all fc flags setup to init routine in all cases.

File: crypto/evp/e_aes.c (hunk 3 above)
Comment (dot-asm) [reply crypto/evp/e_aes.c]: My impression was that you can't have km extensions without km. The chain of checks OPENSSL_s390x_facilities suggests it, i.e. if there is no message-security-assist facility, it won't check for message-security-assist-4 one. But ok, fair enough...

File: crypto/evp/e_aes.c [hunk 5]
```diff
@@ -1056,18 +1107,16 @@ typedef struct {
...
-# define S390X_aes_192_CAPABLE ((OPENSSL_s390xcap_P.km[0] &	\
-                                 S390X_CAPBIT(S390X_AES_192)) &&\
-                                (OPENSSL_s390xcap_P.kmc[0] &	\
-                                 S390X_CAPBIT(S390X_AES_192)))
-# define S390X_aes_256_CAPABLE ((OPENSSL_s390xcap_P.km[0] &	\
-                                 S390X_CAPBIT(S390X_AES_256)) &&\
-                                (OPENSSL_s390xcap_P.kmc[0] &	\
-                                 S390X_CAPBIT(S390X_AES_256)))
+/* Convert key size to function code. */
+# define S390X_AES_FC(keylen)  (S390X_AES_128 + ((((keylen) << 3) - 128) >> 6))
```
Comment (dot-asm) [inline crypto/evp/e_aes.c]: I'd like to suggest one last thing, could you add `[16,24,32] -> [18,19,20]` to commentary line above? I reckon it would be right moment to squash things to your liking.

File: crypto/evp/e_aes.c (hunk 3 above)
Comment (p-steuer) [reply crypto/evp/e_aes.c]: My understanding of the feature detection architecture is the following:

The facility bits stored by stfle instruction indicate the availability of msa and msa-extensions.

Dependence between msa-extensions is the following (from current POP 1-18 to 1-22): msa-ex1, msa-ex2, msa-ex3 (bit 76), and msa-ex6 require the msa (bit 17). msa-ex4 (bit 77) requires msa-ex3. msa-ex5 (bit 57) requires msa-ex2's sha-512 capabilities. msa-ex7 requires msa-ex5. POP does not list any dependencies for msa-ex8 (bit 146) in the cited chapters. However, on 4-101 it says "Bit 76 is one when bit 146 is one.", i.e., msa-ex8 requires msa-ex3. ("a requires b" in this context means, that availability of a implies availability of b).

msa levels which introduce new intructions have a corresponding facility bit, those which only add new function codes to instructions dont. A msa level's facility bit guarantees the availability of at least function code 0 (query) for the intruction(s) introduced on that level.

Bottom line is that msa extensions dont seem to be "strictly inclusive" in the sense that availability of a msa-extension level implies to availability of all smaller msa-extension levels. However, i havent seen a real machine that implements non-inclusive msa levels so i think its okay to stop msa level detection once the first not installed level is detected. Another thing is that i would not make any assumptions about an instruction's installed function codes without checking query/function code 0 return value. In general i consider its always save to (1) check stfle return value for msa indications, (2) check the instructions' query (which are available with the detected msa levels) for function codes.

File: crypto/evp/e_aes.c (hunk 5 above)
Comment (p-steuer) [reply crypto/evp/e_aes.c]: ok

**Comment by p-steuer** [comment]
Squashed and force pushed.

**Comment by dot-asm** [comment]
Since it formally needs re-approval I'm leaving "pending 2n review" label.

File: crypto/evp/e_aes.c (hunk 3 above)
Comment (dot-asm) [reply crypto/evp/e_aes.c]: > availability of a msa-extension level implies to availability of all smaller msa-extension levels.

I'm just looking at OPENSSL_s390x_facilities, and it's arranged as if there is such relationship. If there is no base message-security-assist detected, it does not look for *any* message-security-assist. And if there is no 4 it doesn't look for 5, and if there is no 5 it doesn't look for 8... Well, something to ponder over later on...

Comment (p-steuer) [reply crypto/evp/e_aes.c]: Yes, im aware of the OPENSSL_s390x_facilities code, but as i said, though the architecture seems to allow "non-inclusive" msa levels (i cant find any statement supporting inclusive levels and listing dependencies between levels explicitly is an indication against it), i dont think there are actual implementations (apart from virtualization) of the architecture which use that freedom.

**Comment by dot-asm** [comment]
Merged. Thanks!
